src/Algebra/Properties/CancellativeCommutativeSemiring/GCD.agda

```diff
@@ -0,0 +1,183 @@
+-- Re-exporting definition of GCD, primality, properties of GCD in semiring
+
+open import Algebra.Divisibility _≈_ _*_ public
+  using (_∣_; _∣∣_)
+open import Algebra.Primality _≈_ _*_ 0# 1# public
+  using (Coprime)
+open import Algebra.Properties.Semiring.GCD semiring public
+
+------------------------------------------------------------------------
+-- Properties of GCD
+
+x≈0⊎y≉0⇒Coprime[x/gcd,y/gcd] :
+  ∀ {x y d} (isGCD : IsGCD x y d) → (x ≉ 0#) ⊎ (y ≉ 0#) →
+  let open IsGCD isGCD
+  in  Coprime quot₁ quot₂     -- x/gcd(x,y) is coprime with y/gcd(x,y)
+                              -- if any of x, y is nonzero
+
+x≈0⊎y≉0⇒Coprime[x/gcd,y/gcd] {x} {y} {d} isGCD nz⊎nz {c}
+                             (q₁ , q₁c≈quot₁) (q₂ , q₂c≈quot₂) = c∣1
+  where
+  open IsGCD isGCD using (greatest; quot₁; quot₂; quot₁∙gcd≈x; quot₂∙gcd≈y)
+  d≉0     = x≉0⊎y≉0⇒gcd≉0 isGCD nz⊎nz
+  q₁*dc≈x = begin
+    q₁ * (d * c)    ≈⟨ Of*CSemig.x∙yz≈xz∙y q₁ d c ⟩
+    (q₁ * c) * d    ≈⟨ *-congʳ q₁c≈quot₁ ⟩
+    quot₁ * d       ≈⟨ quot₁∙gcd≈x ⟩
+    x               ∎
+
+  q₂*dc≈y = begin
+    q₂ * (d * c)    ≈⟨ Of*CSemig.x∙yz≈xz∙y q₂ d c ⟩
+    (q₂ * c) * d    ≈⟨ *-congʳ q₂c≈quot₂ ⟩
+    quot₂ * d       ≈⟨ quot₂∙gcd≈y ⟩
+    y               ∎
+
+  dc∣x = q₁ , q₁*dc≈x
+  dc∣y = q₂ , q₂*dc≈y
+  c∣1  = let
+           (q , q*dc≈d) = greatest dc∣x dc∣y
+           d*qc≈d*1     = begin
+             d * (q * c)   ≈⟨ Of*CSemig.x∙yz≈y∙xz d q c ⟩
+             q * (d * c)   ≈⟨ q*dc≈d ⟩
+             d             ≈⟨ sym (*-identityʳ d) ⟩
+             d * 1#        ∎
+
+           qc≈1 = *-cancelˡ-nonZero {d} (q * c) 1# d≉0 d*qc≈d*1
+        in
+        q , qc≈1
+
+------------------------------------------------------------------------------
+-- Recall that  x ∣∣ y  means that x and y are equal as association classes by
+-- division. For example, for ℕ  this means  x ≡ y,
+-- for ℤ  this means (x ≡ y  or  x ≡ -y).
+
+GCD-unique : ∀ {x x' y y' d d'} → x ∣∣ x' → y ∣∣ y' →
+             IsGCD x y d → IsGCD x' y' d' → d ∣∣ d'
+                                            -- gcd-s for assoc-equal pairs are assoc-equal.
+GCD-unique {x} {x'} {y} {y'} {d} {d'} (x∣x' , x'∣x) (y∣y' , y'∣y) isGCD isGCD' =
+  d∣d' , d'∣d
+  where
+  open IsGCD isGCD  using () renaming (divides₁ to d∣x; divides₂ to d∣y)
+  open IsGCD isGCD' using () renaming (divides₁ to d'∣x'; divides₂ to d'∣y')
+  d∣x' = ∣-trans d∣x x∣x'
+  d∣y' = ∣-trans d∣y y∣y'
+  d∣d' = IsGCD.greatest isGCD' d∣x' d∣y'
+  d'∣x = ∣-trans d'∣x' x'∣x
+  d'∣y = ∣-trans d'∣y' y'∣y
+  d'∣d = IsGCD.greatest isGCD d'∣x d'∣y
+
+------------------------------------------------------------------------------
+-- gcd-distr  is an important lemma of the gcd distributivity:
+-- gcd (c*a) (c*b)  assoc-equal  c * (gcd a b).
+
+gcd-distr : ∀ {a b c d d'} → IsGCD a b d → IsGCD (c * a) (c * b) d' → d' ∣∣ (c * d)
+gcd-distr {a} {b} {c} {d} {d'} isGCD-a-b-d isGCD-ca-cb-d' =  aux (c ≟ 0#)
```

Comment:
This proof is pretty messy and unreadable. There must be a way of breaking it up into meaningful constituent parts, in the same way we have done so in `Data.Nat.GCD`!

Reply:
This is not so simple!
I do not know of how to simplify it any essentially.
Data.Nat.GCD  would not help. Because it uses the properties of division with remainder (``_/_`` relies on division with remainder), on the properties of ``_/_`` on `ℕ`.
And generic  CancellativeCommutativeSemiring  does not possess such an operation with such properties. 
Among ``R : CancellativeCommutativeSemiring`` there are some domains R for which there is not possible a division operation with the properties of Nat.DivMod... In particular there is not possible in general to solve  ``x ∣? y``  in R,  
while for `ℕ` it is solved via divMod.

gcd-distr  is a subtle and a particular lemma. It nicety is in that its proof does not use anything besides the three
axioms of GCD (and the properties of generic ``_∣_``).
The first impression is that  gcd-distr  cannot be derived only from these laws. But it occurs that the goal is achieved by combining these three laws repeatedly, and in a certain way.

I do not find this lemma in textbooks. This is a folklore.  A certain referee has pointed at its formulation. 
Then, I searched for the question at a mathematical forum, and found there a proof sketch.
The true author of the lemma is not known. It is almost the same as the generic Euclid's lemma (which is also 
a folklore). 
I have rewritten this proof in detail, with generalizing it from IntegralDomain to  CancellativeCommutativeSemiring,
and programmed it.
Hear is this short and clear in-formal proof:

[proof.lagda.zip](https://github.com/agda/agda-stdlib/files/5771220/proof.lagda.zip)

Can you provide a simpler proof?

In Agda, it takes some code. Thus, the line
```
If c = 0 then ``gcd ac bc  ∥  gcd 0 0  ∣∣  0  ∣∣  c ∙ (gcd a b)``.
```
of 65 characters is expressed in Agda in 240 characters (not counting trailing blanks).

This is, for example,  due to the following reasons.
a) Agda needs all details.
b) For readability, these assignments with '=' need to be one per line.
c) In comparison to Nat,  `≡` is replaced in many cases with ``∣∣``, which is expanded to the expressions
    containing two pairs of values, the two equalities to be proved, the two factor values to be introduced.  

Please, consider the issue once more.

Reply:
Here's an example of one of your earlier proofs that I've just rewritten from:
```agda
x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd] : ∀ {x y d} → (x ≉ 0#) ⊎ (y ≉ 0#) →
                               ((isGCDᶜ (q₁ , _) (q₂ , _) _) : IsGCD x y d) →
                               Coprime q₁ q₂     
x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd] {x} {y} {d} nz⊎nz
  isGCD@(isGCDᶜ (quot₁ , quot₁∙gcd≈x) (quot₂ , quot₂∙gcd≈y) greatest)
  {c} (q₁ , q₁c≈quot₁) (q₂ , q₂c≈quot₂) = c∣1
  where
  d≉0     = x≉0∨y≉0⇒gcd≉0 isGCD nz⊎nz
  q₁*dc≈x = begin
    q₁ * (d * c)    ≈⟨ Of*CSemig.x∙yz≈xz∙y q₁ d c ⟩
    (q₁ * c) * d    ≈⟨ *-congʳ q₁c≈quot₁ ⟩
    quot₁ * d       ≈⟨ quot₁∙gcd≈x ⟩
    x               ∎

  q₂*dc≈y = begin
    q₂ * (d * c)    ≈⟨ Of*CSemig.x∙yz≈xz∙y q₂ d c ⟩
    (q₂ * c) * d    ≈⟨ *-congʳ q₂c≈quot₂ ⟩
    quot₂ * d       ≈⟨ quot₂∙gcd≈y ⟩
    y               ∎

  dc∣x = q₁ , q₁*dc≈x
  dc∣y = q₂ , q₂*dc≈y
  c∣1  = let
           (q , q*dc≈d) = greatest dc∣x dc∣y
           d*qc≈d*1     = begin
             d * (q * c)   ≈⟨ Of*CSemig.x∙yz≈y∙xz d q c ⟩
             q * (d * c)   ≈⟨ q*dc≈d ⟩
             d             ≈⟨ sym (*-identityʳ d) ⟩
             d * 1#        ∎

           qc≈1 = *-cancelˡ-nonZero {d} (q * c) 1# d≉0 d*qc≈d*1
        in
        q , qc≈1
```
to
```agda
x∣y∧z∣x/y⇒x*z∣y : ∀ {x y z} → ((x/y , _) : x ∣ y) → z ∣ x/y → x * z ∣ y 
x∣y∧z∣x/y⇒x*z∣y {x} {y} {z} (x/y , x/y*x≈y) (p , p*z≈x/y) = p , (begin
  p * (x * z)  ≈⟨ Of*CSemig.x∙yz≈xz∙y p x z ⟩
  (p * z) * x  ≈⟨ *-congʳ p*z≈x/y ⟩
  x/y * x      ≈⟨ x/y*x≈y ⟩
  y            ∎)

x*y∣x⇒y∣1 : ∀ {x y} → x ≉ 0# → x * y ∣ x → y ∣ 1#
x*y∣x⇒y∣1 {x} {y} x≉0 (q , q*xy≈x) = q , *-cancelˡ-nonZero (q * y) 1# x≉0 (begin
  x * (q * y) ≈⟨  Of*CSemig.x∙yz≈y∙xz x q y ⟩
  q * (x * y) ≈⟨  q*xy≈x ⟩
  x           ≈˘⟨ *-identityʳ x ⟩
  x * 1#      ∎)

x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]2 : ∀ {x y d} → x ≉ 0# ⊎ y ≉ 0# →
                                ((isGCDᶜ (q₁ , _) (q₂ , _) _) : IsGCD x y d) →
                                Coprime q₁ q₂
x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]2 x≉0∨y≉0 gcd@(isGCDᶜ d∣x d∣y greatest) x/d∣z y/d∣z =
  x*y∣x⇒y∣1 (x≉0∨y≉0⇒gcd≉0 gcd x≉0∨y≉0) (greatest
    (x∣y∧z∣x/y⇒x*z∣y d∣x x/d∣z)
    (x∣y∧z∣x/y⇒x*z∣y d∣y y/d∣z))
```
Can you see the difference? There's significantly less code duplication and I've split out the sub-results into useful lemmas that can be re-used by other proofs. In contrast your proof is brittle as it relies on implementation details of the divisibility relation, hard to understand and therefore difficult to refactor.

Please do the same for this proof.

Reply:
This is an improvement for ``x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]`` . Thank you.
I must have introduced this ``x∣y∧z∣x/y⇒x*z∣y``.
Also I did not know of the possibility to set things like `` ((foo  , _) : x ∣ y) →`` 
to signatures. 
``x*y∣x⇒y∣1``  is also useful.
Now,  I 
a) rename  `x∣y∧z∣x/y⇒x*z∣y`  to  `x∣y∧z∣x/y⇒xz∣y`
    (in mathematics, they often write `xy` for `x * y`, and I suggest to do this in names
     - - when it is clear that this means   `x * y` or `x  ∙ y`),
b) rename   `x*y∣x⇒y∣1` to `xy∣x⇒y∣1`,
c) move  `x∣y∧z∣x/y⇒xz∣y` to Algebra.Properties.CommutativeSemigroup.Divisibility,
    because it is more generic, it is for CommutativeSemigroup and its `_∙_`.

Reply:
> Putting foo1 = x∣y∧z∣x/y⇒x*z∣y d∣x x/d∣z, who divides who according to foo1 ?
Putting foo2 = x∣y∧z∣x/y⇒x*z∣y d∣y y/d∣z, does it return the same value as foo1 ?

I'm afraid I do not understand what you mean.

> Now it is somewhat simper and shorter. I wonder of what else can be done.

Nowhere should the proof explicitly rely on the fact that divisibility is implemented as a pair. It should never directly construct or deconstruct a divisibility proof. Anything else breaks the abstraction boundary. Please fix.

Reply:
> Nowhere should the proof explicitly rely on the fact that divisibility is implemented as a pair. It should never directly 
> construct or deconstruct a divisibility proof. Anything else breaks the abstraction boundary. Please fix.

It looks like you contradict yourself and also the pattern matching tool propagated by Agda.
For example, in  Algebra.Properties.CancellativeCommutativeSemiring.GCD you write
```
x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd] : ∀ {x y d} → x ≉ 0# ⊎ y ≉ 0# →                           
                                ((mkIsGCD (q₁ , _) (q₂ , _) _) : IsGCD x y d) → Coprime q₁ q₂
```     
Here ``(q₁ , _)`` directly deconstructs the divisibility proof.
Do you require the same for `∃` ? 
Divisibility is defined via `∃`.  And 
``(a , Pa) :   ∃ \x -> P x)``  is all right but  ``(q , q∙a≈b) : a ∣ b)``
is not?
Agda programs (I expect standard library also) are full  with deconstructing and constructing pairs,
and the same is with other records.
Many functions use pattern matching that deconstructs some record, whether it is a pair or some other 
record.
If we define 
```
module _ {α} (A : Set α) (_∙_ : Op₂ A α)  where 
  data _∣_ A → A → Set _
     where
     divides : ∀ {q} → q ∙ x ≈ y → x ∣ y                     (II)
```
will this be good to pattern-match   ``(divides {q} qa≈b) :  a | b``    (II).
Suppose that for the next library version you decide to define  divisibility as (II).
Then a great amount of code will break.
Further, one could decide to change  ``divides`` to ``div``. Again, a great amount of code will break.
Further, one could decide to change there  ``y ≈ q ∙ x`` to  ``q ∙ x ≈ y`` (as you have done).
Again, a great amount of code will break.

The requirement of preserving the user code under changing the program format for the same mathematical definition 
is too strong for a language that uses pattern matching intensively.
For example, I see three naturally looking definitions in Agda for the divisibility notion `_|_`.
None of them is better than others.

I do not understand you.
Am I missing something?

Reply:
>> Nowhere should the proof explicitly rely on the fact that divisibility is implemented as a pair. It should never directly
construct or deconstruct a divisibility proof. Anything else breaks the abstraction boundary. Please fix.
>
> It looks like you contradict yourself and also the pattern matching construct propagated by Agda.
For example, in Algebra.Properties.Semigroup.Divisibility we write (and you silently agreed with this)
∣-trans (p , px≈y) (q , qy≈z) = ....
If we wrote ∣-trans x∣y y∣z = foo where p = proj₁ x∣y ...,
then would you decide that this proj₁ does not destruct divisibility by using that this is a pair?

@mechvel there's a huge difference between using the concrete implementation of divisibility in a proof about divisibility vs using the concrete implementation of divisibility in a proof about greatest common divisors. 

Unfortunately I do not have the time to give you a crash course in the concepts of abstraction and encapsulation via a Github pull request, nor the time to rewrite this proof myself. If you don't understand the difference, then I would recommend that you read up about "Software abstraction".

In the meantime we can either 1) remove this proof from the PR and we can merge the rest in, or 2) put this PR on hold?

Reply:
> here's a huge difference between using the concrete implementation of divisibility in a proof about divisibility vs using 
> the concrete implementation of divisibility in a proof about greatest common divisors.

I am trying to understand you. 
Probably you mean that for a higher level item `Item` (functions, type) it is better not to use an explicit construct for any lower level item `Item'`. 
For example,  gcd-distr is about `_|_` and IsGCD. A higher level IsGCD is expressed via a lower level   `_|_` .
And it is desirable for a proof for any property P of IsGCD not to use that `_|_` is expressed as a pair.
Rather it needs to use lemmas in which `_|_` takes part as whole. For example, 
 ``lemma : Foo1 x (x | y) -> Foo2  x y (x | z)``.  Here `lemma` is a separate function considered as of a lower level,
and its  implementation has a bit more rights to use the explicit construction for ` _|_`.
Then, if another program version uses a different  construct for  `_|_`, some higher level parts of the program would not need to be changed.  The wider is applied this approach the more stable fragments has the program.
Also reasoning expressed via statements like ``Foo1 x (x | y) -> Foo2  x y (x | z)`` looks more clear than reasoning that uses pattern matching of  ``x | y`` against a pair. 

Is this what you mean? Is this a "concept of abstraction and encapsulation" ?

May be a common-reason people write "abstract-encapsulated programs" without knowing this?
This is like writing and speaking prose without knowing of a concept of prose and without looking into a crash course about this concept.
 
Another point. Probably it is not true that totally following this requirement always leads to a better program. Is it? 
For example, you succeeded with it in the implementation of  `x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]`,
but  its signature pattern-matches divisibility against a pair:  
``... ((mkIsGCD (q₁ , _) (q₂ , _) _) : IsGCD x y d) → Coprime q₁ q₂``.

Probably, there are many functions that cannot be splited into the abstract-encapsulated form in a reasonable and nice way. This is also a matter of taste.
Currently I do not see how to further "abstract-encapsulate" the function  `gcd-distr`.
May be I shall find this later. 
But my impression is that this requirement needs not be applied totally and bureaucratically.
For example,  gcd-distr has a clear signature (without breaking "abstraction"), its proof is not large and is type-checked.
If we merge it and the generic GCD project to standard, this will be a great improvement. This is more important than  whether some proof is more abstracted or less. Anyway, someone could replace it later with a nicer proof.

Reply:
> Is this what you mean? Is this a "concept of abstraction and encapsulation" ?

Yes.

> May be a common-reason people write "abstract-encapsulated programs" without knowing this?
This is like writing and speaking prose without knowing of a concept of prose and without looking into a crash course about this concept.

I'm afraid I don't understand what you mean here.

> Probably, there are many functions that cannot be splited into the abstract-encapsulated form in a reasonable and nice way.

Yes this is true. For example you say...

> For example, you succeeded with it in the implementation of x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd],
but its signature pattern-matches divisibility against a pair:
... ((mkIsGCD (q₁ , _) (q₂ , _) _) : IsGCD x y d) → Coprime q₁ q₂.

In the case of `x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]` it is impossible to avoid breaking the abstraction in the type as it refers to one of the implementation details. Notice that the abstraction remains in place in the proof.

However `gcd-distr` does not require you to break the abstraction either in the type or in the proof. It is entirely possible to rewrite in the same way as I showed you how to rewrite the proof of `x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]`.

> There is is also a matter of taste in this. ... But my impression is that this requirement needs not be applied totally and bureaucratically.

I respectfully disagree, writing clean, maintainable and reusable code that doesn't hinder further development is not a matter of taste but good software design!

> For example, gcd-distr has a clear signature (without breaking "abstraction"), its proof is not large and is type-checked.

At the moment this proof is _huge_ by the standards of the library. I think it's easily in the top 5 largest proofs we have, and it's possibly the 2nd largest! That's no small feat in a ~100,000 lines of code.

> If we merge it and the generic GCD project to standard, this will be a great improvement. This is more important than whether some proof is more abstracted or less. Anyway, someone could replace it later with a nicer proof.

Unfortunately I'm not a fan of this course of action. If everyone did this, then the standard library would soon end up completely unmaintainable. It's also not reasonable to expect other people to come along and fix code later on. It is simply accruing technical debt.

> Currently I do not see how to further "abstract-encapsulate" the function gcd-distr.
May be I shall find this later. ... 

That's fine that you don't understand how to do it.  All that means is that this isn't a suitable contribution to the standard library at the moment, apologies.

Reply:
>> Probably, there are many functions that cannot be splited into the abstract-encapsulated form in a 
>> reasonable and nice way.

> Yes this is true. For example you say...

Another example: 
introduce to the module where _|_ is defined the functions
`quotient` and `divEquality` that extract the parts of the data of  ``(x | y)``.
Then, instead of writing   ``f = let (q , q*a==b) = a|b in <use q and q*a==b`` 
 it is possible to write
``` 
q = quotient a|b;   q*a==b = divEquality a|b   
<use q  and  q*a==b>
```
Does this satisfies the approach of abstraction and encapsulating?

Reply:
Partially, because you're right you could then change the definition of `_|_` underneath without breaking the code. But it would still be the wrong level of abstraction to use in this proof, because you should be able to prove the properties of GCD simply from building up via lemmas about `_||_` and `_|_`.

Reply:
In mathematical textbooks a discourse like
"Suppose ``a | b``. Then, let `q` be the corresponding quotient, so that ``q * a == b``. 
Hence ``q | (b * c)``, and so on ...
"
is a common place. There are similar fragments longer than my gcd-distr code.
The thing is natural, no reader would think of "abstraction and encapsulation" at this point.  
Anyway, if it is rewritten to your style of  ``gcd-uni x≉0⊎y≉0⇒Coprime[x/gcd,y/gcd]``,  then it would be better  for the library   as well as for reading.
I am going to consider this after a certain delay.

Reply:
If you head over to the Lean Zulip, you'd see that if there's one thing that the mathematicians have learned about formalizing things in Lean's mathlib is... APIs and the value of encapsulation!  What's going in mathematics is that math texts set up a "local context" or scope, where some definitions and equations are valid and visible, and then use them freely. So they use this local scope mechanism instead of setting up an explicit record to bundle up this information.

In other words, CS has a few things to teach to math about how to be precise about what the mathematical vernacular in common use actually means.

Reply:
1. The whole subject is evident, and has nothing to do with mathematics.
2. If we define divisibility as a record `Divides` with constructor `divides`, the fields `quotient` and
`equality`, then ``(divides q eq) : Divides a b`` is an abstract enough thing.
3. Probably Matthew is right in that gcd-distr can be written in a better way.
4. Even the current version of divisibility, IsGCD,  and gcd-distr is still much better than the whole design in lib-1.5 for divisibility, GCD, Rational and such - even if the latter's are programmed very "encapsulatedly". 
Generally it is desirable to distinguish intuitively important things from not so important.

Reply:
More point: 
when a language relies heavily on pattern matching, this drives it from the abstraction you are so eager of. 
For example, I expect to find in standard library programs similar to 
```
f (_ :: (_ , y) :: ps) = y + ( f  ps)
f  _ = 0
```
What do you write for the "abstract & encapsulated" style? 
Is it
```
f ps  = case null? ps
           of \
           { (yes _) -> 0 
           ; (no notNull-ps) -> 
                   case null? (tail ps notNull-ps)
                   of \
                    { (yes _) -> 0
                    ; (no notNull-tlPs) -> proj-2  (head (tail ps notNull-ps) notNull-tlPs) + (f ...)
                    } 
              }
```
?

Reply:
That second is not at all "abstract and encapsulated". I would call that imperative-style programming.  It is an ugly version of the first code, with pretty much nothing going for it.

Relying heavily on pattern-matching makes things representation-dependent. That is fine for implementing the base routines, and then gets problematic, eventually, for higher-level code.

The code for `f` clearly is a kind of fold that works on even length lists. The "abstract and encapsulated" way of doing this would be to provide such a fold abstraction, and then instantiate it with `0` and `\x acc -> proj-2 x + acc`.

Reply:
> That second is not at all "abstract and encapsulated".

Is it representation dependent? I think, not. 
It uses only  1)  `null?` to decide of emptiness,  
2) `head` and `tail` for a nonempty  a list, these actually constitute a definition of what is a list, 
3) second projection for a direct product.
Only basic mathematical notions, the representation details are not visible.  
?
> to provide such a fold abstraction, and then instantiate it with 0 and \x acc -> proj-2 x + acc.

There is a concrete example. Can you write the function source?

Reply:
> and then gets problematic, eventually, for higher-level code.

Operating with  ``gcd , IsGCD`` is higher-level.
Does this mean that one needs to avoid writing things like
```
let (mkIsGCD g g|a g|b greatest) = gcd a b in ...
```
?